Connecting to Gubernator on Kubernetes

We want to use Gubernator on our project and have implemented it with local builds using docker-compose. Now we are trying to set up a gubernator cluster on our dev Kubernetes environment to test it out further.

We are running into some issues though where we get connection refused on our http calls to Gubernator

Logs from our ingest service which tries to connect to gubernator
```
14:34:23.375 ERROR i.o.u.service.http.HttpErrorHandler - Unexpected error for POST /api/v3/data/ingest/int-oblx-monitor?timestampPrecision=milliseconds
io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: gubernator-http.gubernator.svc.cluster.local/10.100.135.70:9080                                                                                             
Caused by: java.net.ConnectException: Connection refused at java.base/sun.nio.ch.Net.pollConnect(Native Method)
```

Logs from our Gubernator cluster, currently only 1 node active to make debugging easier
```
time="2021-01-12T14:18:40Z" level=info msg="Gubernator 1.0.0-rc.3 (amd64/linux)"
time="2021-01-12T14:18:40Z" level=debug msg="Debug enabled"
time="2021-01-12T14:18:40Z" level=debug msg="K8s peer pool config found"
time="2021-01-12T14:18:40Z" level=debug msg="(gubernator.DaemonConfig) {\n GRPCListenAddress: (string) (len=12) \"localhost:81\",\n HTTPListenAddress: (string) (len=12) \"localhost:80\",\n AdvertiseAddress: (string) (len=12) \"localhost:81\",\n CacheSize: (int) 50000,\n Behaviors: (gubernator.BehaviorConfig) {\n  BatchTimeout: (time.Duration) 0s,\n  BatchWait: (time.Duration) 0s,\n  BatchLimit: (int) 0,\n  GlobalSyncWait: (time.Duration) 0s,\n  GlobalTimeout: (time.Duration) 0s,\n  GlobalBatchLimit: (int) 0,\n  MultiRegionSyncWait: (time.Duration) 0s,\n  MultiRegionTimeout: (time.Duration) 0s,\n  MultiRegionBatchLimit: (int) 0\n },\n DataCenter: (string) \"\",\n PeerDiscoveryType: (string) (len=3) \"k8s\",\n EtcdPoolConf: (gubernator.EtcdPoolConfig) {\n  AdvertiseAddress: (string) (len=12) \"localhost:81\",\n  Client: (*clientv3.Client)(<nil>),\n  OnUpdate: (gubernator.UpdateFunc) <nil>,\n  KeyPrefix: (string) (len=17) \"/gubernator-peers\",\n  EtcdConfig: (*clientv3.Config)(0xc00029c420)({\n   Endpoints: ([]string) (len=1 cap=1) {\n    (string) (len=14) \"localhost:2379\"\n   },\n   AutoSyncInterval: (time.Duration) 0s,\n   DialTimeout: (time.Duration) 5s,\n   DialKeepAliveTime: (time.Duration) 0s,\n   DialKeepAliveTimeout: (time.Duration) 0s,\n   MaxCallSendMsgSize: (int) 0,\n   MaxCallRecvMsgSize: (int) 0,\n   TLS: (*tls.Config)(<nil>),\n   Username: (string) \"\",\n   Password: (string) \"\",\n   RejectOldCluster: (bool) false,\n   DialOptions: ([]grpc.DialOption) <nil>,\n   LogConfig: (*zap.Config)(<nil>),\n   Context: (context.Context) <nil>,\n   PermitWithoutStream: (bool) false\n  }),\n  Logger: (logrus.FieldLogger) <nil>,\n  DataCenter: (string) \"\"\n },\n K8PoolConf: (gubernator.K8sPoolConfig) {\n  Logger: (logrus.FieldLogger) <nil>,\n  OnUpdate: (gubernator.UpdateFunc) <nil>,\n  Namespace: (string) (len=10) \"gubernator\",\n  Selector: (string) (len=14) \"app=gubernator\",\n  PodIP: (string) (len=10) \"10.244.2.8\",\n  PodPort: (string) (len=2) \"81\"\n },\n MemberListPoolConf: (gubernator.MemberListPoolConfig) {\n  MemberListAddress: (string) (len=14) \"localhost:7946\",\n  AdvertiseAddress: (string) (len=12) \"localhost:81\",\n  KnownNodes: ([]string) {\n  },\n  OnUpdate: (gubernator.UpdateFunc) <nil>,\n  NodeName: (string) \"\",\n  Logger: (logrus.FieldLogger) <nil>,\n  DataCenter: (string) \"\"\n },\n Picker: (gubernator.PeerPicker) <nil>,\n Logger: (logrus.FieldLogger) <nil>,\n TLS: (*gubernator.TLSConfig)(<nil>)\n}\n"
time="2021-01-12T14:18:40Z" level=info msg="GRPC Listening on localhost:81 ..." category=gubernator
time="2021-01-12T14:18:40Z" level=debug msg="Queue (Add) 'gubernator/gubernator-http' - <nil>" category=gubernator
time="2021-01-12T14:18:40Z" level=debug msg="Queue (Add) 'gubernator/gubernator' - <nil>" category=gubernator
time="2021-01-12T14:18:40Z" level=info msg="HTTP Gateway Listening on localhost:80 ..." category=gubernator
time="2021-01-12T14:18:41Z" level=debug msg="Queue (Update) 'gubernator/gubernator-http' - <nil>" category=gubernator
time="2021-01-12T14:18:41Z" level=debug msg="Fetching peer list from endpoints API" category=gubernator
time="2021-01-12T14:18:41Z" level=debug msg="Peer: {DataCenter: HTTPAddress: GRPCAddress:10.244.2.8:81 IsOwner:true}\n" category=gubernator
time="2021-01-12T14:18:41Z" level=debug msg="Peer: {DataCenter: HTTPAddress: GRPCAddress:10.244.2.8:81 IsOwner:true}\n" category=gubernator
time="2021-01-12T14:18:41Z" level=debug msg="peers updated" category=gubernator peers="[{  10.244.2.8:81 true} {  10.244.2.8:81 true}]"
time="2021-01-12T14:18:41Z" level=debug msg="Queue (Update) 'gubernator/gubernator' - <nil>" category=gubernator
```

We deployed using the default `k8s-deployment.yaml` from the repository, we then added a non-headless service as well as the standard headless one and used that one for the URL but still the same results.

**Thoughts**

- Is it possible that the Http Server is not started correctly/bound to correct address, logs say it is listening on `localhost:80` but in further debug log messages like this HTTPAdress is blank like this: `time="2021-01-12T13:39:04Z" level=debug msg="Peer: {DataCenter: HTTPAddress: GRPCAddress:10.244.3.110:81 IsOwner:false}\n" category=gubernator`
- Do we need to uncomment the port description in the default k8s-deployment file? Have tried both with and without but no difference so far

Our current k8s-deployment.yaml which includes the extra service
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: gubernator
  namespace: gubernator
  labels:
    app: gubernator
spec:
  replicas: 1
  selector:
    matchLabels:
      app: gubernator
  template:
    metadata:
      labels:
        app: gubernator
    spec:
      serviceAccountName: gubernator
      containers:
        - image: thrawn01/gubernator:latest
          imagePullPolicy: IfNotPresent
          ports:
            - name: grpc-port
              containerPort: 81
            - name: http-port
              containerPort: 80
          name: gubernator
          env:
          - name: GUBER_K8S_NAMESPACE
            valueFrom:
              fieldRef:
                fieldPath: metadata.namespace
          - name: GUBER_K8S_POD_IP
            valueFrom:
              fieldRef:
                fieldPath: status.podIP
          # Use the k8s API for peer discovery
          - name: GUBER_PEER_DISCOVERY_TYPE
            value: "k8s"
          # This should match the port number GRPC is listening on
          # as defined by `containerPort`
          - name: GUBER_K8S_POD_PORT
            value: "81"
          # The selector used when listing endpoints. This selector
          # should only select gubernator peers.
          - name: GUBER_K8S_ENDPOINTS_SELECTOR
            value: "app=gubernator"
          # Enable debug for diagnosing issues
          - name: GUBER_DEBUG
            value: "true"
      restartPolicy: Always
---
apiVersion: v1
kind: Service
metadata:
  name: gubernator
  labels:
    app: gubernator
spec:
  clusterIP: None
  # ports:
  # - name: grpc-port
  #   targetPort: 81
  #   protocol: TCP
  #   port: 81
  # - name: http-port
  #   targetPort: 80
  #   protocol: TCP
  #   port: 80
  selector:
    app: gubernator
---
apiVersion: v1
kind: Service
metadata:
  name: gubernator-http
  labels:
    app: gubernator
spec:
  ports:
  - name: http-port
    targetPort: http-port
    protocol: TCP
    port: 9080
  selector:
    app: gubernator
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: gubernator
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: get-endpoints
rules:
- apiGroups:
  - ""
  resources:
  - endpoints
  verbs:
  - list
  - watch
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: get-endpoints
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: get-endpoints
subjects:
- kind: ServiceAccount
  name: gubernator
```

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -4,6 +4,11 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [1.0.0-rc.6] - 2020-12-21
+## Changes
+* Update the k8s example to reflect the latest changes from the release
+candidate.
+
 ## [1.0.0-rc.5] - 2020-12-21
 ## Changes
 * Respect SIGTERM from docker during shutdown

--- a/k8s-deployment.yaml
+++ b/k8s-deployment.yaml
@@ -33,6 +33,12 @@ spec:
             valueFrom:
               fieldRef:
                 fieldPath: status.podIP
+          # Must set the GRPC and HTTP addresses, as gubernator
+          # defaults to listening on localhost only
+          - name: GUBER_GRPC_ADDRESS
+            value: 0.0.0.0:81
+          - name: GUBER_HTTP_ADDRESS
+            value: 0.0.0.0:80
           # Use the k8s API for peer discovery
           - name: GUBER_PEER_DISCOVERY_TYPE
             value: "k8s"
@@ -44,6 +50,10 @@ spec:
           # should only select gubernator peers.
           - name: GUBER_K8S_ENDPOINTS_SELECTOR
             value: "app=gubernator"
+          # Gubernator can watch 'endpoints' for changes to the peers
+          # or it can watch 'pods' (Defaults to 'endpoints')
+          # - name: GUBER_K8S_WATCH_MECHANISM
+          #  value: "endpoints"
           # Enable debug for diagnosing issues
           - name: GUBER_DEBUG
             value: "true"

